spectator-api/src/main/java/com/netflix/spectator/api/patterns/MonotonicCounter.java

```diff
@@ -32,72 +32,135 @@
+     * executor will be used which is limited to a single thread to minimize the worst
+     * case resource usage for collecting the counter data. Use a custom executor if more
+     * resources are needed or if the polling operation is expensive.
+     *
+     * @return
+     *     This builder instance to allow chaining of operations.
+     */
+    public Builder withExecutor(ScheduledExecutorService executor) {
+      this.executor = executor;
+      return this;
+    }
+
+    /**
+     * Poll the value of the provided {@link Number}. The implementation provided must
+     * be thread safe. The most common usages of this are to monitor instances of
+     * {@link java.util.concurrent.atomic.AtomicInteger},
+     * {@link java.util.concurrent.atomic.AtomicLong}, or
+     * {@link java.util.concurrent.atomic.LongAdder}.
+     *
+     * @param number
+     *     Thread-safe implementation of {@link Number} used to access the value.
+     * @return
+     *     The number that was passed in to allow the builder to be used inline as part
+     *     of an assignment.
+     */
+    public <T extends Number> T monitor(T number) {
+      return monitor(number, Number::longValue);
+    }
+
+    /**
+     * Poll by executing {@code f(obj)} and reporting the returned value. The provided
+     * function must be thread safe and cheap to execute. Expensive operations, including
+     * any IO or network calls, should not be performed inline unless using a custom
+     * executor by calling {@link #withExecutor(ScheduledExecutorService)}. Assume that the
+     * function will be called frequently and may be called concurrently.
+     *
+     * A weak reference will be kept to {@code obj} so that monitoring the object will
+     * not prevent garbage collection. The gauge will go away when {@code obj} is collected.
+     * To explicitly disable polling call {@link #remove(Id)} with the same id used with
+     * this builder.
+     *
+     * @param obj
+     *     Object used to compute a value.
+     * @param f
+     *     Function that is applied on the value for the number.
+     * @return
+     *     The object that was passed in so the registration can be done as part of an assignment
+     *     statement.
+     */
+    @SuppressWarnings("unchecked")
+    public <T> T monitor(T obj, ToLongFunction<T> f) {
+      final Id id = baseId.withTags(extraTags);
+      final Counter counter = registry.counter(id);
+      final Tuple<T> tuple = new Tuple<>(counter, obj, f);
+
+      ConcurrentMap<Id, Object> state = registry.state();
+      Object c = Utils.computeIfAbsent(state, id, i -> tuple);
+      if (!(c instanceof Tuple)) {
+        Utils.propagateTypeError(registry, id, MonotonicCounter.class, c.getClass());
+      } else {
+        Tuple<T> t = (Tuple<T>) c;
+        t.schedule(registry, executor);
+      }
+
+      return obj;
     }
-    return value;
   }
 
   /** Keep track of the object reference, counter, and other associated bookkeeping info. */
```

Comment:
`Tuple` does not connote the semantics of this type to me.  Consider something like `Bookkeeper`, `DeltaCalculator`, or something else more descriptive.